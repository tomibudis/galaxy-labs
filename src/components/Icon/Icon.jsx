import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { SIZE } from '#constants';

import styles from './styles.module.scss';

const path = 'https://raw.githubusercontent.com/twbs/icons/main/icons/';
const format = '.svg';
export const ICON_TESTID = 'icon-testid';

const Icon = ({ name, className, size }) => {
  const source = useMemo(() => {
    return path + name + format;
  }, [name]);

  return (
    <img
      data-testid={ICON_TESTID}
      src={source}
      className={cx(className, {
        [styles.iconSm]: size === SIZE.SM,
        [styles.iconMd]: size === SIZE.MD,
        [styles.iconLg]: size === SIZE.LG,
      })}
    />
  );
};

Icon.propTypes = {
  /** Name of icon */
  name: PropTypes.string.isRequired,
  /** Additional classname */
  className: PropTypes.string,
  /** Size of icon */
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
};

Icon.defaultProps = {
  size: 'sm',
};

export default Icon;
