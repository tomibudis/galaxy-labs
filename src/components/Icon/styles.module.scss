.iconSm {
  width: 16px;
}

.iconMd {
  width: 22px;
}

.iconLg {
  width: 30px;
}