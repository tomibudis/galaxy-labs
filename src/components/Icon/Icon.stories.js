import React from 'react';
import Icon from '.';

export default {
  title: 'Components/Icon',
  component: Icon,
  argTypes: {
    children: {
      table: {
        disable: true,
      },
    },
  },
};

const Template = (args) => {
  return (
    <div style={{ width: 600 }}>
      <Icon {...args} className="mx-1" size="lg" />
      <Icon {...args} className="mx-1" size="md" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
      <Icon {...args} className="mx-1" />
    </div>
  );
};

export const TypeIcon = Template.bind({});

TypeIcon.args = {
  name: 'youtube',
};
