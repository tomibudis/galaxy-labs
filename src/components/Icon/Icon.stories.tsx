import React from 'react';

import { Meta, ComponentStory } from '@storybook/react';

import Icon from '.';
import icons from '../../constants/icons.json';

export default {
  title: 'Components/Icon',
  component: Icon,
  argTypes: {
    children: {
      table: {
        disable: true,
      },
    },
    name: {
      table: {
        disable: true,
      },
    },
  },
} as Meta;

const renderIcon = (name, size): JSX.Element => {
  return (
    <div
      className="d-flex flex-column justify-content-center m-2"
      style={{ width: '100px' }}
    >
      <Icon className="mx-auto" name={name} size={size} />
      <span className="text-center mt-2">{name}</span>
    </div>
  );
};

const Template: ComponentStory<typeof Icon> = (args) => {
  return (
    <div
      style={{ width: '600px' }}
      className="d-flex flex-wrap justify-content-between align-items-start "
    >
      {icons.map((icon) => {
        return renderIcon(icon, args.size);
      })}
    </div>
  );
};

export const TypeIcon = Template.bind({});
TypeIcon.args = {
  size: 'lg',
};
