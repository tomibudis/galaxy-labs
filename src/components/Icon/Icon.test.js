import React from 'react';
import { cleanup, render } from '@testing-library/react';

import Icon, { ICON_TESTID } from './Icon';

afterEach(() => {
  cleanup();
});

const IconComponent = (props) => {
  return <Icon {...props} name="youtube" />;
};

describe('<Icon />', () => {
  it('render component', () => {
    const { getByTestId } = render(<IconComponent />);
    expect(getByTestId(ICON_TESTID)).toBeTruthy();
  });

  test.each([
    ['sm', '.iconSm'],
    ['md', '.iconMd'],
    ['lg', '.iconLg'],
  ])('render %s icon', (size, expected) => {
    const { baseElement } = render(<IconComponent size={size} />);
    expect(baseElement.querySelector(expected)).toBeTruthy();
  });

  it('add additional classname', () => {
    const { baseElement } = render(<IconComponent className="mx-auto" />);
    expect(baseElement.querySelector('.mx-auto')).toBeTruthy();
  });

  it('matched with snapshot file', () => {
    const { baseElement } = render(<IconComponent />);
    expect(baseElement).toMatchSnapshot();
  });
});
