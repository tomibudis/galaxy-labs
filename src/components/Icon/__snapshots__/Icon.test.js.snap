// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> matched with snapshot file 1`] = `
<body>
  <div>
    <img
      class="iconSm"
      data-testid="icon-testid"
      src="https://raw.githubusercontent.com/twbs/icons/main/icons/youtube.svg"
    />
  </div>
</body>
`;
